huaweicloud/services/dc/resource_huaweicloud_dc_virtual_interface.go

```diff
@@ -0,0 +1,458 @@
+	_, err := interfaces.Update(client, interfaceId, opts)
+	if err != nil {
+		return fmt.Errorf("error updating virtual interface (%s): %s", interfaceId, err)
+	}
+	return nil
+}
+
+func updateVirtualInterfaceBasicInfo(client *golangsdk.ServiceClient, d *schema.ResourceData) error {
+	var (
+		interfaceId = d.Id()
+
+		opts = interfaces.UpdateOpts{
+			Name:          d.Get("name").(string),
+			Description:   utils.String(d.Get("description").(string)),
+			Bandwidth:     d.Get("bandwidth").(int),
+			RemoteEpGroup: utils.ExpandToStringList(d.Get("remote_ep_group").([]interface{})),
+		}
+	)
+
+	return updateVirtualInterface(client, interfaceId, opts)
+}
+
+func closeVirtualInterfaceNetworkDetection(client *golangsdk.ServiceClient, d *schema.ResourceData) error {
+	var (
+		interfaceId = d.Id()
+		opts        = interfaces.UpdateOpts{}
+	)
+
+	// BFD and NQA would not be enabled at the same time.
+	if d.HasChange("enable_bfd") && !d.Get("enable_bfd").(bool) {
+		opts.EnableBfd = utils.Bool(false)
+	} else if d.HasChange("enable_nqa") && !d.Get("enable_nqa").(bool) {
+		opts.EnableNqa = utils.Bool(false)
+	}
+	if reflect.DeepEqual(opts, interfaces.UpdateOpts{}) {
+		return nil
+	}
+
+	_, err := interfaces.Update(client, interfaceId, opts)
+	if err != nil {
+		return fmt.Errorf("error closing network detection of the virtual interface (%s): %s", interfaceId, err)
+	}
+	return nil
+}
+
+func openVirtualInterfaceNetworkDetection(client *golangsdk.ServiceClient, d *schema.ResourceData) error {
+	var (
+		interfaceId     = d.Id()
+		detectionOpened = false
+		opts            = interfaces.UpdateOpts{}
+	)
+
+	if d.HasChange("enable_bfd") && d.Get("enable_bfd").(bool) {
+		detectionOpened = true
+		opts.EnableBfd = utils.Bool(true)
+	}
+	if d.HasChange("enable_nqa") && d.Get("enable_nqa").(bool) {
+		if detectionOpened {
+			return fmt.Errorf("BFD and NQA cannot be enabled at the same time")
+		}
+		opts.EnableNqa = utils.Bool(true)
+	}
+	if reflect.DeepEqual(opts, interfaces.UpdateOpts{}) {
+		return nil
+	}
+
+	_, err := interfaces.Update(client, interfaceId, opts)
+	if err != nil {
+		return fmt.Errorf("error opening network detection of the virtual interface (%s): %s", interfaceId, err)
+	}
+	return nil
+}
+
+func resourceVirtualInterfaceUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	cfg := meta.(*config.Config)
+	client, err := cfg.DcV3Client(cfg.GetRegion(d))
+	if err != nil {
+		return diag.Errorf("error creating DC v3 client: %s", err)
+	}
+
+	if d.HasChanges("name", "description", "bandwidth", "remote_ep_group") {
+		if err = updateVirtualInterfaceBasicInfo(client, d); err != nil {
+			return diag.FromErr(err)
+		}
+	}
+	if d.HasChanges("enable_bfd", "enable_nqa") {
+		if err = closeVirtualInterfaceNetworkDetection(client, d); err != nil {
+			return diag.FromErr(err)
+		}
+		if err = openVirtualInterfaceNetworkDetection(client, d); err != nil {
```

Comment:
it‘s very confusing here. 
maybe you can return two opts 
then invoke interfaces.Update(client, interfaceId, opts)

Reply:
Added some comments to describe these functions call.